libraries/protocol/include/scorum/protocol/betting/market.hpp

```diff
@@ -2,16 +2,188 @@
+    : public base_market_type<base_market_without_parameters_type<Kind, Wincases...>, Kind, Wincases...>
+{
+    base_market_without_parameters_type() = default;
+
+    template <typename WincasePair> auto create_wincase(WincasePair&&) const
+    {
+        typename WincasePair::left_wincase result{};
+        return result;
+    }
+
+    bool less(const base_market_without_parameters_type<Kind, Wincases...>&) const
+    {
+        return false;
+    }
+};
+
+template <market_kind Kind, typename... Wincases>
+struct base_market_with_threshold_type
+    : public base_market_type<base_market_with_threshold_type<Kind, Wincases...>, Kind, Wincases...>
+{
+    threshold_type::value_type threshold = 0;
+
+    base_market_with_threshold_type() = default;
+
+    base_market_with_threshold_type(const threshold_type::value_type threshold_)
+        : threshold(threshold_)
+    {
+    }
+
+    template <typename WincasePair> auto create_wincase(WincasePair&&) const
+    {
+        typename WincasePair::left_wincase result{ threshold };
+        return result;
+    }
+
+    bool less(const base_market_with_threshold_type<Kind, Wincases...>& other) const
+    {
+        return threshold < other.threshold;
+    }
+};
+
+template <market_kind Kind, typename... Wincases>
+struct base_market_with_score_type
+    : public base_market_type<base_market_with_score_type<Kind, Wincases...>, Kind, Wincases...>
 {
-    market_kind kind;
-    fc::flat_set<wincase_pair> wincases;
+    uint16_t home = 0;
+    uint16_t away = 0;
+
+    base_market_with_score_type() = default;
+
+    base_market_with_score_type(const int16_t home_, const int16_t away_)
+        : home(home_)
+        , away(away_)
+    {
+    }
+
+    template <typename WincasePair> auto create_wincase(WincasePair&&) const
+    {
+        typename WincasePair::left_wincase result{ home, away };
+        return result;
+    }
+
+    bool less(const base_market_with_score_type<Kind, Wincases...>& other) const
+    {
+        return home < other.home || (!(other.home < home) && away < other.away);
+    }
 };
+
+// it can be set separately
+using result_home_market
+    = base_market_without_parameters_type<market_kind::result, WINCASE(result_home, result_draw_away)>;
+
+using result_draw_market
+    = base_market_without_parameters_type<market_kind::result, WINCASE(result_draw, result_home_away)>;
+
+using result_away_market
+    = base_market_without_parameters_type<market_kind::result, WINCASE(result_away, result_home_draw)>;
+
+using round_market = base_market_without_parameters_type<market_kind::round, WINCASE(round_home, round_away)>;
+
+using handicap_market
+    = base_market_with_threshold_type<market_kind::handicap, WINCASE(handicap_home_over, handicap_home_under)>;
+
+// or grouped together
+using correct_score_market
+    = base_market_without_parameters_type<market_kind::correct_score,
+                                          WINCASE(correct_score_home_yes, correct_score_home_no),
```

Comment:
Why not to use template here instead of macro definition? It will improve navigation across the codebase in the IDE